eisbuk-admin/functions/src/utils.ts

```diff
@@ -186,3 +186,75 @@ export const sendRequest = <
     }
     req.end();
   });
+
+type ErrorCode = functions.https.FunctionsErrorCode;
+
+/** */
+export class EisbukHttpsError extends functions.https.HttpsError {
+  /**
+   * A wrapper around `functions.https.HttpsError` used to log errors
+   * to the functions console before returning gRPC error to the client
+   * @param code `functions.https.FunctionsErrorCode`
+   * @param message error message
+   * @param details error details
+   */
+  constructor(code: ErrorCode, message: string, details?: unknown) {
+    functions.logger.error(code, message, details);
+    super(code, message, details);
+  }
+}
+
+/**
+ * A function wrapper used to enforce a timeout on async
+ * operations with possibly long exectution period
+ * @param fn an async function we want to execute within timeout boundary
+ * @returns `fn`'s resolved value
+ */
+export const runWithTimeout = async <T extends any>(
+  fn: () => Promise<T>,
+  { timeout } = { timeout: 5000 }
+): Promise<T> => {
+  // set a timeout boundry for a function
+  const timeoutBoundary = setTimeout(() => {
+    throw new functions.https.HttpsError("aborted", HTTPSErrors.TimedOut);
+  }, timeout);
+
+  const res = await fn();
+
+  // clear timeout function (the `fn` has resolved within the timeout boundry)
+  clearTimeout(timeoutBoundary);
+
+  return res;
+};
+
+/**
+ * Checks payload (or any other record structure) for required fields.
+ * Throws error if one or more of the fields aren't present
+ * @param payload record to validate
+ * @param requiredFields array of field names (keys)
+ * @param errorMessage (optional) customer error message on failure, falls back to
+ * `HTTPSErrors.MissingParameter`
+ */
+export const checkRequiredFields = (
+  payload: Record<string, any> | undefined | null,
+  requiredFields: string[],
+  errorMessage: string = HTTPSErrors.MissingParameter
```

Comment:
This parameter is currently never used. 
It would be nice to see the missing fields enumerated in the error message.
I think not having the message as a parameter would make for a better dev experience. The function should build an error message that states what fields were missing.

Reply:
I left it like this in case we want to pass a different error to the same checking logic. One possible use case I had in mind was when checking template data (which should be stored in firestore) so we would still pass template data object and required fields into the `checkRequiredFields`, but would return a different error: not saying that some parameters are missing, but stating that some tamplate data is missing from admin setup (not from request payload)

Reply:
I think discussing the merits of the extra parameters for future usage would drive the conversation away from the main point. In my opinion we should have that discussion when the extra parameter comes into use. I think it would be better to introduce it when we start using it, but I won't put my foot down about this.

My point is that the `message` an exception carries should hold all the most relevant information regarding what error occurs. I believe that's very important. In some contexts only the message will be available. I will put my foot down about this.

As things stand, as a (dev) user of the `checkRequiredFields` function, I can't get it to output a message that summarizes what went wrong. `checkRequiredFields`, you had one job!

Again: I'm slightly against coding functionalities that will be used later (I fear that, lacking an actual use case, some problems might be missed), but what bothers me the most is that we don't get the missing field names in the message, and I believe we should.

If we don't include the missing fields in the error message we should have a strong reason not to do so. I fail to see such reason.